File: backend/app/services/reset_token_service.py
import hashlib
import secrets
from datetime import datetime, timedelta

from flask import current_app

from app.extensions import db
from app.models.password_reset_token import PasswordResetToken

RESET_TTL_SECONDS = 900  # 15 minutes
INVITE_TTL_SECONDS = 86400  # 24 hours


def _hash_token(raw_token: str) -> str:
    return hashlib.sha256(raw_token.encode()).hexdigest()


def _expires_at(ttl: int) -> datetime:
    return datetime.utcnow() + timedelta(seconds=ttl)


def _purge_expired_for_user(user_id: str) -> None:
    now = datetime.utcnow()
    PasswordResetToken.query.filter(
        PasswordResetToken.user_id == user_id,
        PasswordResetToken.expires_at <= now,
    ).delete(synchronize_session=False)


def _store_reset(user_id: str, token_hash: str, ttl: int) -> None:
    _purge_expired_for_user(user_id)
    PasswordResetToken.query.filter_by(user_id=user_id).delete(synchronize_session=False)
    db.session.add(
        PasswordResetToken(
            token_hash=token_hash,
            user_id=user_id,
            expires_at=_expires_at(ttl),
        )
    )
    db.session.commit()


def _get_reset(token_hash: str) -> str | None:
    row = PasswordResetToken.query.get(token_hash)
    if not row:
        return None
    if datetime.utcnow() > row.expires_at:
        db.session.delete(row)
        db.session.commit()
        return None
    return str(row.user_id)


def generate_reset_token() -> tuple[str, str]:
    raw = secrets.token_urlsafe(32)
    return raw, _hash_token(raw)


def store_reset_token(user_id: str, token_hash: str, ttl: int = RESET_TTL_SECONDS) -> None:
    _store_reset(user_id, token_hash, ttl)


def store_invite_token(user_id: str, token_hash: str) -> None:
    _store_reset(user_id, token_hash, INVITE_TTL_SECONDS)


def get_user_id_for_token(raw_token: str) -> str | None:
    return _get_reset(_hash_token(raw_token))


def delete_reset_token(raw_token: str) -> None:
    token_hash = _hash_token(raw_token)
    row = PasswordResetToken.query.get(token_hash)
    if row:
        db.session.delete(row)
        db.session.commit()


def build_reset_url(raw_token: str, *, invite: bool = False) -> str:
    frontend = current_app.config.get("FRONTEND_URL", "http://localhost:5173")
    url = f"{frontend}/reset-password?token={raw_token}"
    if invite:
        url += "&invite=1"
    return url

File: server/src/services/resetTokenService.ts
import { createHash, randomBytes } from 'node:crypto'
import { and, eq, lte } from 'drizzle-orm'
import { config } from '../config.js'
import { db } from '../db/index.js'
import { passwordResetTokens } from '../db/schema/index.js'

const RESET_TTL_SECONDS = 900
const INVITE_TTL_SECONDS = 86400

function hashToken(rawToken: string): string {
  return createHash('sha256').update(rawToken).digest('hex')
}

function expiresAt(ttl: number): Date {
  return new Date(Date.now() + ttl * 1000)
}

async function purgeExpiredForUser(userId: string): Promise<void> {
  const now = new Date()
  await db
    .delete(passwordResetTokens)
    .where(and(eq(passwordResetTokens.userId, userId), lte(passwordResetTokens.expiresAt, now)))
}

async function storeReset(userId: string, tokenHash: string, ttl: number): Promise<void> {
  await purgeExpiredForUser(userId)
  await db.delete(passwordResetTokens).where(eq(passwordResetTokens.userId, userId))
  await db.insert(passwordResetTokens).values({
    tokenHash,
    userId,
    expiresAt: expiresAt(ttl),
  })
}

async function getReset(tokenHash: string): Promise<string | null> {
  const [row] = await db
    .select()
    .from(passwordResetTokens)
    .where(eq(passwordResetTokens.tokenHash, tokenHash))
    .limit(1)
  if (!row) return null
  if (new Date() > row.expiresAt) {
    await db.delete(passwordResetTokens).where(eq(passwordResetTokens.tokenHash, tokenHash))
    return null
  }
  return row.userId
}

export function generateResetToken(): [string, string] {
  const raw = randomBytes(32).toString('base64url')
  return [raw, hashToken(raw)]
}

export async function storeResetToken(userId: string, tokenHash: string, ttl = RESET_TTL_SECONDS): Promise<void> {
  await storeReset(userId, tokenHash, ttl)
}

export async function storeInviteToken(userId: string, tokenHash: string): Promise<void> {
  await storeReset(userId, tokenHash, INVITE_TTL_SECONDS)
}

export async function getUserIdForToken(rawToken: string): Promise<string | null> {
  return getReset(hashToken(rawToken))
}

export async function deleteResetToken(rawToken: string): Promise<void> {
  const tokenHash = hashToken(rawToken)
  await db.delete(passwordResetTokens).where(eq(passwordResetTokens.tokenHash, tokenHash))
}

export function buildResetUrl(rawToken: string, invite = false): string {
  let url = `${config.frontendUrl}/reset-password?token=${rawToken}`
  if (invite) url += '&invite=1'
  return url
}
